### PR TITLE
Use of non-API entry points in `data.table`

### LINKED ISSUE #55: blog about C API
@aitap since you are familiar with the C API issue https://github.com/Rdatatable/data.table/issues/6180
I was wondering if you may be interested to write a blog for the Raft about the current state of the R C API and how data.table uses it? If so, could you please write a general / gentle introduction, so that even people who are not C API experts (like me) could understand.
Target readers would be any current/future data.table developers, as well as other R package developers using the C API.

### PR DESCRIPTION
**Pull request by aitap**
This is not yet ready to be posted — the unfinished pieces are marked by `<!-- TODO: ... -->` in the document source — but if you have any suggestions for the overall shape of the post or any edits that can be applied right away, I'll be glad to hear them.

**Edit**: is there a list of potentially applicable categories?

For the image, I can't decide between [a still from a certain MC Hammer video](https://github.com/user-attachments/assets/c93d477c-56d8-46bd-9106-340c102b2e96) ([source](https://www.youtube.com/watch?v=otCpCn0l4Wo)) and a [more visual metaphor](https://github.com/user-attachments/assets/64bed2b7-686d-41a7-b702-dd444f1047b7) for something that one does fully knowing what they are doing, at their own peril, until one day it fails ([source](https://en.wikipedia.org/wiki/File:Suicide_cable.jpg), CC-BY-SA-4.0).

Fixes: #55

### CODE REVIEW

**Comment by aitap** [comment]
Dear @tdhock,

Here's the preliminary version of the post. I think it contains or at least links to all the necessary information. If its form needs improvements or its content needs additions, I'll be glad to perform the necessary changes. Can we ask Michael Chirico for a reality check? I've done a deep dive into the R API, but I've never tuned a hash table manually. Also, my pre-R-3.1 history of both data.table and R may be very lacking.

If we manage to get rid of all the NOTEs regarding non-API entry points, how about submitting an article to the R journal, presenting `data.table` as a case study in API compliance, with everyone who did the planning and implementation as co-authors?

**Comment by tdhock** [comment]
thanks! these are the current categories on https://rdatatable-community.github.io/The-Raft/

ambassadors (2)
announcements (7)
application package (1)
benchmarks (1)
bridge package (2)
community (3)
conferences (1)
developer (4)
documentation (2)
extension package (1)
funding opportunity (2)
governance (2)
grant (9)
guest post (2)
opinion (1)
partner package (1)
performance (1)
releases (1)
seal of approval (6)
testing (2)
tips (4)
translation (1)
travel (1)
tutorials (4

File: posts/2024-12-12-non-api-use/index.qmd [hunk 1]
```diff
@@ -0,0 +1,1276 @@
...
+Since the correctness of the modification operation hinges on the
+reference count being 0 (and it may be important whether it's exactly 1
+or above 1), the same amount of _useful_ information can be conveyed by
+printing `MAYBE_REFERENCED()` and `MAYBE_SHARED()` instead of `NAMED()`:
+
+```c
+if (verbose) {
+  Rprintf(_("RHS for item %d has been duplicated because MAYBE_REFERENCED==%d MAYBE_SHARED==%d, but then is being plonked. length(values)==%d; length(cols)==%d)\n"),
+          i+1, MAYBE_REFERENCED(thisvalue), MAYBE_SHARED(thisvalue), length(values), length(cols));
+            // ^^^^^^^^^^^^^^^^ API function
```
Comment (tdhock) [inline posts/2024-12-12-non-api-use/index.qmd:421]: is there a PR for this suggested change?

I see a link to a PR for another issue above, so I assume there is none for this one? so should somebody submit a PR?

more generally, it would be useful to have a comment/sentence at the end about the current status of each issue, so we can know if more work still needs to be done.

File: posts/2024-12-12-non-api-use/index.qmd [hunk 2]
```diff
@@ -0,0 +1,1276 @@
...
+  table][datatable_chmatch_lookup], [cleaning up before
+  exit][datatable_chmatch_cleanup2]
+* `src/fmelt.c`: [combining factor levels by merging their `CHARSXP`s in
+  a common array with indices in `TRUELENGTH`][datatable_fmelt_truelen]
+
+Since there does't seem to be any intent to allow using R API to place
+arbitrary integer values inside unused `SEXP` fields, `data.table` will
+have to look up the `CHARSXP` values using the externally available
+information. Performing $O(nk)$ direct pointer comparisons would scale
+poorly, so for an $O(1)$ individual lookup `data.table` could build a
```
Comment (tdhock) [inline posts/2024-12-12-non-api-use/index.qmd]: I like the big O notation here

Comment (tdhock) [reply posts/2024-12-12-non-api-use/index.qmd]: we could use atime to verify the asymptotic performance if we make this change

**Comment by tdhock** [comment]
this is really excellent overall thanks very much.
I guess you could probably submit it for peer review  @ R journal as well?

I wonder if you can think of any easy/logical way to break it up into several differnt blog pages? I did not see one, so I think it is ok to keep it as one big one.

File: posts/2024-12-12-non-api-use/index.qmd [hunk 3]
```diff
@@ -0,0 +1,1276 @@
...
+// returns a const SEXP * pointer, which prevents accidental writes
+```
+
+See also: [PR18775].
+
+Reading the reference counts: `NAMED` {#NAMED}
+-------------------------------------
+
+In plain R, all value types -- numbers, strings, lists -- have
+pass-by-value semantics. Wihout dark and disturbing things in play, such
```
Comment (Anirban166) [inline posts/2024-12-12-non-api-use/index.qmd]: Without

File: posts/2024-12-12-non-api-use/index.qmd [hunk 4]
```diff
@@ -0,0 +1,1276 @@
...
+ * [marking][gp_for_calling] condition handlers as "calling" (executing
+   on top of where the condition was signalled in the call stack), as
+   opposed to "non-calling" (executing at the site of the `tryCatch`
+   call)
+ * [marking][gp_for_assignment] objects in complex assignment calls
+ * storing the [S4 object bit][gp_for_s4]
+ * [marking][gp_for_jit] functions as (un)suitable for bytecode
+   compilation
+ * [marking][gp_for_growable] vectors as growable
+ * [marking][gp_for_missing] provided ("actial") function arguments as
```
Comment (Anirban166) [inline posts/2024-12-12-non-api-use/index.qmd]: actual?

File: posts/2024-12-12-non-api-use/index.qmd [hunk 5]
```diff
@@ -0,0 +1,1276 @@
...
+making the change breaks a number of unit tests.
+
+`SETLENGTH`, `SET_GROWABLE_BIT`, `(SET_)TRUELENGTH`
+---------------------------------------------------
+
+### Growable vectors
+
+Since `data.frame`s and `data.table`s are lists, and lists in R are
+value types with <a href="#NAMED">pass-by-value semantics</a>, adding or
+removing a column to one the normally involves allocating a new list
```
Comment (Anirban166) [inline posts/2024-12-12-non-api-use/index.qmd]: > one the normally

omit 'the'

File: posts/2024-12-12-non-api-use/index.qmd [hunk 6]
```diff
@@ -0,0 +1,1276 @@
...
+* `src/chmatch.c`: [saving the original
+  `TRUELENGTH`s][datatable_chmatch_savetl], [remembering the positions
+  of `CHARSXP`s in the table][datatable_chmatch_settl], [cleaning up on
+  error][datatable_chmatch_cleanup1], [looking up strings in the
+  table][datatable_chmatch_lookup], [cleaning up before
+  exit][datatable_chmatch_cleanup2]
+* `src/fmelt.c`: [combining factor levels by merging their `CHARSXP`s in
+  a common array with indices in `TRUELENGTH`][datatable_fmelt_truelen]
+
+Since there does't seem to be any intent to allow using R API to place
```
Comment (Anirban166) [inline posts/2024-12-12-non-api-use/index.qmd]: doesn't

File: posts/2024-12-12-non-api-use/index.qmd [hunk 7]
```diff
@@ -0,0 +1,1276 @@
...
+  operations evaluated with `dt[..., ..., by=...]`, so they are [marked
+  with a `TRUELENGTH` of -1][datatable_dogroups_setlen-1], and the
+  [marked columns][datatable_dogroups_anyspecialstatic] are later
+  re-created.
+* In `src/utils.c`, columns share memory or are ALTREP must be copied.
+  Memory sharing between columns may lead to confusing results when they
+  are altered by reference, and ALTREP columns cannot have `TRUELENGTH`
+  set. The code uses the same trick as with `CHARSXP` objects: if
+  `TRUELENGTH` is set on an object, accessing the object through a
+  different pointer and seeting `TRUELENGTH` set will prove that the
```
Comment (Anirban166) [inline posts/2024-12-12-non-api-use/index.qmd]: > seeting `TRUELENGTH` set

Maybe rephrase?

**Review by Anirban166 — APPROVED** [review]
Impressive stuff!

File: posts/2024-12-12-non-api-use/index.qmd (hunk 1 above)
Comment (aitap) [reply posts/2024-12-12-non-api-use/index.qmd:421]: This one is [merged](https://github.com/Rdatatable/data.table/pull/6420), and I'll make sure to summarise the status of all the issues.

**Comment by aitap** [comment]
Thank you very much for the comments! I think that a better R Journal article could result from describing the fixes after they are implemented, not only plans for them.

Regarding a logical breakdown, the issues could be split by difficulty: easy (rename a function and possibly add a wrapper); `TRUELENGTH` and friends (ALTREP + pointer hashes, could be split into two); `ATTRIB` and friends (currently not clear how to replace).

**Comment by aitap** [comment]
Can we use [this](https://github.com/user-attachments/assets/64bed2b7-686d-41a7-b702-dd444f1047b7) for the post image? I apologise for my previous indecisiveness. Should I create a new PR for the changes?
